connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestForwardingIntegrationTest.java

```diff
@@ -0,0 +1,256 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.StrictStubs.class)
+@SuppressWarnings("unchecked")
+@Category(IntegrationTest.class)
+public class RestForwardingIntegrationTest {
+
+    private Map<String, Object> sslConfig;
+    @Mock
+    private Plugins plugins;
+    private RestClient followerClient;
+    private RestServer followerServer;
+    @Mock
+    private Herder followerHerder;
+    private RestClient leaderClient;
+    private RestServer leaderServer;
+    @Mock
+    private Herder leaderHerder;
+
+    private SslContextFactory factory;
+    private CloseableHttpClient httpClient;
+    private Collection<CloseableHttpResponse> responses = new ArrayList<>();
+
+    @Before
+    public void setUp() throws IOException, GeneralSecurityException {
+        sslConfig = TestSslUtils.createSslConfig(false, true, Mode.SERVER, TestUtils.tempFile(), "testCert");
+    }
+
+    @After
+    public void tearDown() throws IOException {
+        for (CloseableHttpResponse response: responses) {
+            response.close();
+        }
+        AtomicReference<Throwable> firstException = new AtomicReference<>();
+        Utils.closeAllQuietly(
+                firstException,
+                "clientsAndServers",
+                httpClient,
+                followerServer != null ? followerServer::stop : null,
+                leaderServer != null ? leaderServer::stop : null,
+                factory != null ? factory::stop : null
+        );
+        if (firstException.get() != null) {
+            throw new RuntimeException("Unable to cleanly close resources", firstException.get());
+        }
+    }
+
+    @Test
+    public void testRestForwardNoSsl() throws Exception {
+        // A cluster with no SSL configured whatsoever, using HTTP
+        testRestForwardToLeader(false, false, false);
+    }
+
+    @Test
+    public void testRestForwardNoSslDualListener() throws Exception {
+        // A cluster configured with HTTPS listeners, but still advertising HTTP
+        testRestForwardToLeader(true, false, false);
+    }
+
+    @Test
+    public void testRestForwardLeaderSsl() throws Exception {
+        // A heterogeneous cluster where the leader rolls to advertise HTTPS before the follower
+        testRestForwardToLeader(true, false, true);
+    }
+
+    @Test
+    public void testRestForwardFollowerSsl() throws Exception {
+        // A heterogeneous cluster where the follower rolls to advertise HTTPS before the leader
+        testRestForwardToLeader(true, true, false);
+    }
+    @Test
```

Comment:
Nit: add a newline between test cases

```suggestion
    }

    @Test
```